### DISCUSSION TITLE
Shared Directory not visible in a bare Debian 11 VM?

### DISCUSSION THREAD

**Discussion by yllekz** [utmapp/UTM#3755] [UTM Help (macOS)]
**Describe the issue**
I installed Debian 11 (no desktop environment) and the Spice Guest Tools per [this link](https://mac.getutm.app/support/). I then added a shared directory and restarted the VM. After looking all over the disk in the VM, I cannot locate where this shared directory is on the disk and it appears as if it has no awareness of this setting. Is this not compatible with a headless Debian VM?

**Configuration**
* UTM Version: 3.0.4
* OS Version: 11.6.3
* Intel or Apple Silicon? Apple M1

**Crash log**
No crash

**Debug log**
Attached
```
Running:  -L /Applications/UTM.app/Contents/Resources/qemu -S -qmp tcp:127.0.0.1:4000,server,nowait -nodefaults -vga none -spice "unix=on,addr=/Users/me/Library/Group Containers/WDNLXAD4W8.com.utmapp.UTM/4C9C7828-51CC-4A5E-8F5A-5A4031641267.spice,disable-ticketing=on,image-compression=off,playback-compression=off,streaming-video=off,gl=off" -device virtio-ramfb -device virtio-rng-pci -cpu host -smp cpus=4,sockets=1,cores=4,threads=1 -machine virt,highmem=off -accel hvf -accel tcg,tb-size=1024 -drive if=pflash,format=raw,unit=0,file=/Applications/UTM.app/Contents/Resources/qemu/edk2-aarch64-code.fd,readonly=on -drive if=pflash,format=raw,unit=1,file=/Users/me/Library/Containers/com.utmapp.UTM/Data/Documents/Debian.utm/Images/efi_vars.fd -boot menu=on -m 4096 -device intel-hda -device hda-duplex -name Debian -device nec-usb-xhci,id=usb-bus -device usb-tablet,bus=usb-bus.0 -device usb-mouse,bus=usb-bus.0 -device usb-kbd,bus=usb-bus.0 -device ich9-usb-ehci1,id=usb-controller-0 -device ich9-usb-uhci1,masterbus=usb-controller-0.0,firstport=0,multifunction=on -device ich9-usb-uhci2,masterbus=usb-controller-0.0,firstport=2,multifunction=on -device ich9-usb-uhci3,masterbus=usb-controller-0.0,firstport=4,multifunction=on -chardev spicevmc,name=usbredir,id=usbredirchardev0 -device usb-redir,chardev=usbredirchardev0,id=usbredirdev0,bus=usb-controller-0.0 -chardev spicevmc,name=usbredir,id=usbredirchardev1 -device usb-redir,chardev=usbredirchardev1,id=usbredirdev1,bus=usb-controller-0.0 -chardev spicevmc,name=usbredir,id=usbredirchardev2 -device usb-redir,chardev=usbredirchardev2,id=usbredirdev2,bus=usb-controller-0.0 -device virtio-blk-pci,drive=drive0,bootindex=0 -drive if=none,media=disk,id=drive0,file=/Users/me/Library/Containers/com.utmapp.UTM/Data/Documents/Debian.utm/Images/data.qcow2,cache=writethrough -device virtio-net-pci,mac=8E:7C:FF:BE:5F:D4,netdev=net0 -netdev vmnet-macos,mode=bridged,id=net0 -device virtio-serial -device virtserialport,chardev=vdagent,name=com.redhat.spice.0 -chardev spicevmc,id=vdagent,debug=0,name=vdagent -device virtserialport,chardev=charchannel1,id=channel1,name=org.spice-space.webdav.0 -chardev spiceport,name=org.spice-space.webdav.0,id=charchannel1 -uuid 4C9C7828-51CC-4A5E-8F5A-5A4031641267 -rtc base=localtime
qemu-aarch64-softmmu: -netdev vmnet-macos,mode=bridged,id=net0: info: Started vmnet interface with configuration:
qemu-aarch64-softmmu: -netdev vmnet-macos,mode=bridged,id=net0: info: MTU:              1500
qemu-aarch64-softmmu: -netdev vmnet-macos,mode=bridged,id=net0: info: Max packet size:  1514
qemu-aarch64-softmmu: -netdev vmnet-macos,mode=bridged,id=net0: info: MAC:              <redacted>
qemu-aarch64-softmmu: -netdev vmnet-macos,mode=bridged,id=net0: info: UUID:             38194EB0-52FA-400B-ABD8-67F80D581192
qemu-aarch64-softmmu: warning: Spice: playback:0 (0x11e012940): setsockopt failed, Operation not supported on socket
qemu-aarch64-softmmu: warning: Spice: record:0 (0x11e0129f0): setsockopt failed, Operation not supported on socket
```

**Comment by osy** [comment]
You need spice-webdavd and once it is running you should be able to access a webdav server on localhost. Check documentation for spice-webdavd for more details.

**Comment by samiraguiar** [comment]
Here's what worked for me.

Install the needed tools (make sure to enable unprivileged access when prompted):
```bash
sudo apt install spice-vdagent spice-webdavd davfs2
```

(not sure if restarting is needed)

Shared directory should now be accessible via HTTP, you can check via:
```bash
$ curl http://127.0.0.1:9843/
```

It should list the contents of your share.

To mount it to a directory:
```bash
$ sudo mkdir /mnt/shared
# empty user and pass here
$ sudo mount -t davfs -o noexec http://127.0.0.1:9843/ /mnt/shared/
```

Store the credentials so it doesn't prompt you next time:
```bash
# "anyuser" and "anypass" can really be anything,
# but seems to be required when specifying in the file.
$ sudo cat >> /etc/davfs2/secrets << EOF
# mounted UTM directory
/mnt/shared anyuser anypass
EOF
```

To have it mounted automatically on boot (make sure that creds are stored as above):
```bash
# add to fstab
$ sudo cat >> /etc/fstab << EOF
# mounted UTM directory
http://127.0.0.1:9843/ /mnt/shared davfs _netdev,user 0 0
EOF
```

Reboot the VM with `sudo systemctl reboot` and the mount should be there when the system boots.

**Reply by yllekz** [reply]
Thanks! Does UTM support more than one shared directory? I'm looking to mount a photos folder as well as a separate folder containing Docker app/config data.

**Reply by kasperk81** [reply]
took about half hour to find this answer.
_this_ should be in the document. the gui specific documentation for ubuntu desktop (https://support.cci.drexel.edu/software-courses/utm-virtualization-m1-macs/utm-sharing-folder-linux-guest/) does not help anywhere else... while your answer helps in majority of linux, including ubuntu headless and desktop.

**Reply by preyes323** [reply]
This worked me as well.

**Reply by phlbnks** [reply]
Worked for me too @samiraguiar - thanks a lot

Please add this to the official docs!

**Reply by mkkumar1988** [reply]
@samiraguiar I am getting "qemu-aarch64-softmmu: -qmp tcp:127.0.0.1:4000,server,nowait: Failed to find  available port: Address already in use" error when starting the VMs after following your steps. Please can you gave me any solution.

Thanks

**Reply by samiraguiar** [reply]
@mkkumar1988 you probably have some program listening on port 4000 on the host. Perhaps other QEMU instances, try to kill/shutdown all of your VMs (or even restarting your device) before retrying.

**Reply by osy** [reply]
@mkkumar1988 You're using an outdated version of UTM. Upgrade to the latest 3.x release.

**Reply by Averyyy** [reply]
Hi samiraguiar:
Thanks for your solution and I successfully mounted it. However, I couldn't execute any file complied from a .c file even if I set the mounting option to be "exec". Could you please help me with this?

**Reply by mattpopovich** [reply]
Anyone else get a `lost+found` folder in their shared folder in Linux (Ubuntu) when they mount their shared folder? It's owned by root and I can't delete it no matter how hard I try... (sudo, rmdir, find, lsattr)

**Reply by grr** [reply]
This worked for me, but note that there is bug which will prevent some files with URI special characters (ex. `[test]`) from displaying. The files will display in the output of `curl http://127.0.0.1:9843/` and in the gui file browser (e..g. Thunar), but will not be displayed in any command-line interaction (`ls`, '`find`, etc.). The likely cause is described here: https://github.com/iterate-ch/cyberduck/issues/13864#issuecomment-1310910993
But I'm not sure where to report this.

**Comment by fiore42** [comment]
None of these suggestions worked for me.

I finally found joy following [these steps](https://docs.getutm.app/guest-support/linux/#virtfs).

Host: MacOS Ventura 13.0 on MacBook Air M2
Guest: Ubuntu 22.04 for ARM
UTM 4.0.9

hope it helps!

**Reply by Irratzo** [reply]
@fiore42 Thank you very much for pointing this out! I had overlooked that the instructions how to set up the different UTM shared folder backends for Linux guests are now actually present [in the official UTM docs](https://docs.getutm.app/guest-support/linux). For my Ubuntu 22.04-based VM (Kubuntu), the [VirtFS instructions](https://docs.getutm.app/guest-support/linux/#virtfs) there worked for me out of the box, no installation required. As the UTM docs [point out](https://docs.getutm.app/guest-support/sharing/directory/), VirtFS is expected to be faster than SPICE WebDAV.

**Reply by Saeven** [reply]
On an M1 host, Ubuntu 20.04 guest (ARM), these instructions only "somewhat" work.  Mounting at the command line works:

    sudo mount -t 9p -o trans=virtio share [mount point] -oversion=9p2000.L

However, adding the equivalent line to /etc/fstab does not.

    share	[mount point]	9p	trans=virtio,version=9p2000.L,rw,_netdev,nofail	0	0

**Comment by jbergandino** [comment]
After installing the Spice tools (`sudo apt install spice-vdagent spice-webdavd`), I needed to shut down the VM and manually change the "Directory Share Mode" from "VirtFS" to "SPICE WebDAV" in the UTM VM settings.

Settings -> Sharing -> Directory Share Mode -> SPICE WebDAV

**Comment by Vartkat** [comment]
for fstab try adding 
9p
9pnet
9pnet_virtio

to /etc/initramfs-tools/modules
then 
$ sudo update-initramfs -u
and reboot

hope this helps

**Comment by q316077570** [comment]
Is there no Android UTM?



---- Replied Message ----
| From | ***@***.***> |
| Date | 12/31/2023 12:43 |
| To | ***@***.***> |
| Cc | ***@***.***> |
| Subject | Re: [utmapp/UTM] Shared Directory not visible in a bare Debian 11 VM? (Discussion #3755) |

This worked for me, but note that there is bug which will prevent some files with URI special characters (ex. [test]) from displaying. The files will display in the output of curl http://127.0.0.1:9843/ and in the gui file browser (e..g. Thunar), but will not be displayed in any command-line interaction (ls, 'find, etc.). The likely cause is described here: iterate-ch/cyberduck#13864 (comment)
But I'm not sure where to report this.
